File: 2022/q07.R
library(dplyr)
library(stringr)

commands <- readLines("q07.in")

get_file_structure <- function(commands) {
  #Initialize
  file_size <- 0
  file_structure <- data.frame()
  #Add root/ for easier parsing
  current_path <- "root/"
  
  for (command in commands) {
    if (command == "$ cd /") next
    #Keep track of current directory
    if (grepl("\\$ cd ", command)) {
      #Get the path
      dir_name <- sub("\\$ cd ", "", command)
      if (dir_name == "..") {
        #Remove the last dir
        current_path <- sub("(?<=/)\\w*/$", "", current_path, perl = T)
      } else {
        current_path <- paste0(current_path, dir_name, "/")
      }
    }
    #Pick up command lines starting with a digit as these are files
    if (grepl("^\\d+", command)) {
      #Get the file sizes (delete everything after the space)
      file_size <- as.numeric(sub(" .*", "", command))
      #Add file size to current_path
      file_structure <- rbind(file_structure, 
                              data.frame(path = current_path, 
                                         size = file_size))
      #Add records to above directories if we are in a sub directory
      #Check depth
      if (stringr::str_count(current_path, "/") - 1 > 0){
        #Create paths for all above directories (parent paths) of current_path
        for (position in head(str_locate_all(current_path, "/")[[1]][, 1], - 1)) {
          #Add file size to parent paths
          file_structure <- rbind(file_structure, 
                                  data.frame(path = substr(current_path, 1, position), 
                                             size = file_size))
        }
      }
    }
  }
  return(file_structure)
}
#Part 1----
#Answer:
get_file_structure(commands) %>%
  group_by(path) %>%
  summarise(size = sum(size)) %>%
  filter(size <= 1e5) %>%
  summarise(size = sum(size))

#Part 2----
path_structure <- 
  get_file_structure(commands) %>%
  group_by(path) %>%
  summarise(size = sum(size)) 

#Calculate space to free up
total_space <- 7e7
needed_space <- 3e7
space_to_free_up <- 
  abs(
    total_space - 
    path_structure %>%
      filter(path == "root/") %>%
      pull(size) - 
    needed_space)

#Answer:
get_file_structure(commands) %>%
  group_by(path) %>%
  summarise(size = sum(size)) %>%
  filter(size >= space_to_free_up) %>%
  arrange(size) %>%
  top_n(-1) %>%
  select(size)

File: 2022/python/q07.py
import re

with open('q07.in') as file:
    src = file.readlines()

# Use list to track current position in the file system
# and to make sure you can add file sizes to parents
current_dir = []
# Use a dict to store file sizes per directory
file_system = {}

for line in src:
    # Throw away garbage
    if line.startswith('$ ls') | line.startswith('dir'):
        continue
    # Directory changes with cd
    elif line.startswith('$ cd'):
        # E.g. $ cd a, a will be the target. In case of $ cd .., pop one value from list.
        target = line.split()[2]
        if target == '..':
            current_dir.pop()
        else:
            current_dir.append(target)
    # Files start with digits
    elif re.match(r"^\d+", line):
        file_size = int(line.split()[0])
        # Use strings as dict keys
        # Add size of files to directory and all parent directories
        # If current dir is /a/b/c, add file_size to /a/b/c, /a/b, /a and /
        for path in ['/'.join(current_dir[0:n+1]) for n in range(0, len(current_dir))]:
            # If the path already exists, add it to the total file_size
            if path in file_system:
                file_system[path] = file_system[path] + file_size
            # If the path does not exist, create it using path as a dict key
            else:
                file_system[path] = file_size
    else:
        raise Exception(f"Unknown unix command: {line}. Please fix the input file or add handling for the command.")
# Calculate and print answer
answer1 = sum([value for value in file_system.values() if value <= 100000])
print(f"answer 1: {answer1}")

# Part two. Set up some variables
total_disk_size = 7e7
needed_space = 3e7
current_free_space = total_disk_size - file_system['/']
space_to_free_up = needed_space - current_free_space
# Filter the dictionary for all directories larger than the needed space
filtered_list = filter(lambda size: size >= space_to_free_up, file_system.values())
# Answer is the first value from the sorted list of values remaining
print(f"answer 2: {sorted(filtered_list)[0]}")
